org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/hierarchy/AbstractCallHierarchyViewPart.java

```diff
@@ -0,0 +1,180 @@
+	public void createPartControl(Composite parent) {
+		Composite layout = createLayout(parent);
+		callHierarchyViewer = createCallHierarchyViewer(layout);
+		locationViewer = createLocationViewer(layout);
+		navigationService.installNavigationSupport(locationViewer);
+		navigationService.installNavigationSupport(callHierarchyViewer);
+		callHierarchyViewer.addSelectionChangedListener(this);
+		addActions(getViewSite().getActionBars().getToolBarManager());
+	}
+
+	protected void addActions(IToolBarManager toolBarManager) {
+		toolBarManager.add(refreshAction);
+		for (SetHierarchyTypeAction setHierarchyTypeAction : setHierarchyTypeActions) {
+			toolBarManager.add(setHierarchyTypeAction);
+		}
+	}
+
+	protected Composite createLayout(Composite parent) {
+		return new SashForm(parent, SWT.NONE);
+	}
+
+	protected TreeViewer createCallHierarchyViewer(Composite parent) {
+		TreeViewer treeViewer = new TreeViewer(parent);
+		treeViewer.getControl().setLayoutData(new GridData(GridData.FILL_BOTH));
+		treeViewer.setUseHashlookup(true);
+		treeViewer.setAutoExpandLevel(2);
+		treeViewer.setLabelProvider(createHierarchyLabelProvider());
+		treeViewer.setContentProvider(createHierarchyContentProvider());
+		return treeViewer;
+	}
+
+	protected IContentProvider createHierarchyContentProvider() {
+		return new HierarchyTreeContentProvider(this);
+	}
+
+	protected abstract IBaseLabelProvider createHierarchyLabelProvider();
+
+	protected TableViewer createLocationViewer(Composite parent) {
+		TableViewer locationViewer = new TableViewer(parent);
+		locationViewer.setContentProvider(new ArrayContentProvider());
+		locationViewer.setLabelProvider(createLocationLabelProvider());
+
+		TableLayout layout = new TableLayout();
+		locationViewer.getTable().setLayout(layout);
+		locationViewer.getTable().setHeaderVisible(true);
+
+		Pair<String, ColumnLayoutData>[] locationColumnDescriptions = getLocationColumnDescriptions();
+		IntStream.range(0, locationColumnDescriptions.length).forEach(index -> {
+			layout.addColumnData(locationColumnDescriptions[index].getValue());
+			createColumn(locationViewer.getTable(), locationColumnDescriptions[index], index);
+		});
+
+		return locationViewer;
+	}
+
+	protected Pair<String, ColumnLayoutData>[] getLocationColumnDescriptions() {
+		return new Pair[] { Pair.of("Line", new ColumnWeightData(60)), Pair.of("Call", new ColumnWeightData(300)) };
+	}
+
+	protected void createColumn(Table table, Pair<String, ColumnLayoutData> columnDescription, int index) {
+		TableColumn column = new TableColumn(table, SWT.NONE, index);
+		column.setResizable(columnDescription.getValue().resizable);
+		column.setText(columnDescription.getKey());
+	}
+
+	protected abstract IBaseLabelProvider createLocationLabelProvider();
+
+	@Override
+	public void setFocus() {
+		callHierarchyViewer.getControl().setFocus();
+	}
+
+	@Override
+	public void selectionChanged(SelectionChangedEvent event) {
+		if (event.getSelectionProvider() == callHierarchyViewer) {
```

Comment:
Sure you don't mean .equals here?

Reply:
It was identity comparison in the original code.